`pkgsCross.loongarch64-linux.stdenv` broken by #238154

### Describe the bug
A clear and concise description of what the bug is.

Building `pkgsCross.loongarch64-linux.stdenv` fails with:

```shell
loongarch64-unknown-linux-gnu-stage-static-gcc> checking for thread model used by GCC... single
loongarch64-unknown-linux-gnu-stage-static-gcc> checking whether assembler supports CFI directives... yes
loongarch64-unknown-linux-gnu-stage-static-gcc> checking for target glibc version... 0.0
loongarch64-unknown-linux-gnu-stage-static-gcc> checking for __attribute__((visibility("hidden")))... yes
loongarch64-unknown-linux-gnu-stage-static-gcc> checking for .cfi_sections .debug_frame... yes
loongarch64-unknown-linux-gnu-stage-static-gcc> checking whether the target assembler supports thread-local storage... yes
loongarch64-unknown-linux-gnu-stage-static-gcc> checking whether the thread-local storage support is from emutls... no
loongarch64-unknown-linux-gnu-stage-static-gcc> checking for init priority support... yes
loongarch64-unknown-linux-gnu-stage-static-gcc> configure: updating cache ./config.cache
loongarch64-unknown-linux-gnu-stage-static-gcc> configure: creating ./config.status
loongarch64-unknown-linux-gnu-stage-static-gcc> config.status: creating Makefile
loongarch64-unknown-linux-gnu-stage-static-gcc> config.status: creating auto-target.h
loongarch64-unknown-linux-gnu-stage-static-gcc> config.status: executing default commands
loongarch64-unknown-linux-gnu-stage-static-gcc> make[1]: Entering directory '/build/build/loongarch64-unknown-linux-gnu/libgcc'
loongarch64-unknown-linux-gnu-stage-static-gcc> # If this is the top-level multilib, build all the other
loongarch64-unknown-linux-gnu-stage-static-gcc> ln -s ../../../gcc-12.3.0/libgcc/enable-execute-stack-empty.c enable-execute-stack.c
loongarch64-unknown-linux-gnu-stage-static-gcc> # multilibs.
loongarch64-unknown-linux-gnu-stage-static-gcc> ln -s ../../../gcc-12.3.0/libgcc/unwind-generic.h unwind.h
loongarch64-unknown-linux-gnu-stage-static-gcc> ln -s ../../../gcc-12.3.0/libgcc/config/loongarch/linux-unwind.h md-unwind-support.h
loongarch64-unknown-linux-gnu-stage-static-gcc> ln -s ../../../gcc-12.3.0/libgcc/config/loongarch/sfp-machine.h sfp-machine.h
loongarch64-unknown-linux-gnu-stage-static-gcc> ln -s ../../../gcc-12.3.0/libgcc/gthr-single.h gthr-default.h
loongarch64-unknown-linux-gnu-stage-static-gcc> DEFINES='' HEADERS='' \
loongarch64-unknown-linux-gnu-stage-static-gcc>         ../../../gcc-12.3.0/libgcc/mkheader.sh > tmp-libgcc_tm.h
loongarch64-unknown-linux-gnu-stage-static-gcc> make[1]: *** No rule to make target '../../../gcc-12.3.0/libgcc/config/loongarch/crti.S', needed by 'crti.o'.  Stop.
loongarch64-unknown-linux-gnu-stage-static-gcc> make[1]: *** Waiting for unfinished jobs....
loongarch64-unknown-linux-gnu-stage-static-gcc> /nix/store/r4vxljid3iq94jp7qvd639sps0fscwy3-bash-5.2-p15/bin/bash ../../../gcc-12.3.0/libgcc/../move-if-change tmp-libgcc_tm.h libgcc_tm.h
loongarch64-unknown-linux-gnu-stage-static-gcc> echo timestamp > libgcc_tm.stamp
loongarch64-unknown-linux-gnu-stage-static-gcc> make[1]: Leaving directory '/build/build/loongarch64-unknown-linux-gnu/libgcc'
loongarch64-unknown-linux-gnu-stage-static-gcc> make: *** [Makefile:12949: all-target-libgcc] Error 2
error: builder for '/nix/store/3jbq4p271jffnz42wwmg24lrdvc9p16p-loongarch64-unknown-linux-gnu-stage-static-gcc-12.3.0.drv' failed with exit code 2;
       last 10 log lines:
       > ln -s ../../../gcc-12.3.0/libgcc/config/loongarch/sfp-machine.h sfp-machine.h
       > ln -s ../../../gcc-12.3.0/libgcc/gthr-single.h gthr-default.h
       > DEFINES='' HEADERS='' \
       >   ../../../gcc-12.3.0/libgcc/mkheader.sh > tmp-libgcc_tm.h
       > make[1]: *** No rule to make target '../../../gcc-12.3.0/libgcc/config/loongarch/crti.S', needed by 'crti.o'.  Stop.
       > make[1]: *** Waiting for unfinished jobs....
       > /nix/store/r4vxljid3iq94jp7qvd639sps0fscwy3-bash-5.2-p15/bin/bash ../../../gcc-12.3.0/libgcc/../move-if-change tmp-libgcc_tm.h libgcc_tm.h
       > echo timestamp > libgcc_tm.stamp
       > make[1]: Leaving directory '/build/build/loongarch64-unknown-linux-gnu/libgcc'
       > make: *** [Makefile:12949: all-target-libgcc] Error 2
       For full logs, run 'nix log /nix/store/3jbq4p271jffnz42wwmg24lrdvc9p16p-loongarch64-unknown-linux-gnu-stage-static-gcc-12.3.0.drv'.
error: 1 dependencies of derivation '/nix/store/yl3kybngy6iha06ns2hsq6m801sg5pnw-loongarch64-unknown-linux-gnu-stage-static-gcc-wrapper-12.3.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/kz5pk6qydxfs3ckf369411pw76ccradr-libgcc-loongarch64-unknown-linux-gnu-12.3.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/aamb9z8gz6bm9pw1b4x0glz7xs722rrg-stdenv-linux.drv' failed to build
error: 1 dependencies of derivation '/nix/store/wlknk98q6fhw8av6ifp80bfhkr4p142q-glibc-loongarch64-unknown-linux-gnu-2.37-8.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dm5najbzspvb9pgd28ri8fsk3ww0bgwc-loongarch64-unknown-linux-gnu-stage-final-gcc-wrapper-12.3.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dhgf7kb07jdm26ad39d045zmqgkcmhq9-stdenv-linux.drv' failed to build
```

### Steps To Reproduce
Steps to reproduce the behavior:
1. Use latest nixpkgs commit 5385752f2a6d68c9d807830ff7693505d9b5f7ab
2. Run `nix build .#pkgCross.loongarch64-linux.stdenv`

### Expected behavior
A clear and concise description of what you expected to happen.

It should not fail.

### Screenshots
If applicable, add screenshots to help explain your problem.

### Additional context
Add any other context about the problem here.

Similar issue: https://github.com/NixOS/nixpkgs/issues/243614.

Adding `--enable-initfini-array` does not work.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@amjoseph-nixpkgs

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.2.16-5-pve, Debian GNU/Linux, 12 (bookworm), nobuild`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.12.0`
 - channels(root): `""`
 - channels(jiegec): `"nixpkgs"`
 - nixpkgs: `/home/jiegec/.nix-defexpr/channels/nixpkgs`
```

--- a/pkgs/development/compilers/gcc/common/libgcc-buildstuff.nix
+++ b/pkgs/development/compilers/gcc/common/libgcc-buildstuff.nix
@@ -31,7 +31,22 @@ let
   #
   SHLIB_LC = lib.optionalString stdenv.targetPlatform.isPower "-mnewlib";
 
-in ''
-    echo 'libgcc.a: ${crtstuff-ofiles}' >> libgcc/Makefile.in
-    echo 'SHLIB_LC=${SHLIB_LC}' >> libgcc/Makefile.in
-  ''
+in
+''
+  echo 'libgcc.a: ${crtstuff-ofiles}' >> libgcc/Makefile.in
+  echo 'SHLIB_LC=${SHLIB_LC}' >> libgcc/Makefile.in
+''
+
+  # Meanwhile, crt{i,n}.S are not present on certain platforms
+  # (e.g. LoongArch64), resulting in the following error:
+  #
+  # No rule to make target '../../../gcc-xx.x.x/libgcc/config/loongarch/crti.S', needed by 'crti.o'.  Stop.
+  #
+  # For LoongArch64, a hacky workaround is to simply touch them,
+  # as the platform forces .init_array support.
+  #
+  # https://www.openwall.com/lists/musl/2022/11/09/3
+  #
+  + lib.optionalString stdenv.targetPlatform.isLoongArch64 ''
+  touch libgcc/config/loongarch/crt{i,n}.S
+''

